single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,749 @@
+        # Add in any new IPs to the replica set. Relation handlers require a reference to
+        # a unit.
+        self.on_relation_changed()
+
+        # make sure all nodes in the replica set have the same priority for re-election. This is
+        # necessary in the case that pre-upgrade hook fails to reset the priority of election for
+        # cluster nodes.
+        self.mongo_manager.set_election_priority(priority=1)
+
+    def update_hosts(self):
+        """Update the replica set hosts and remove any unremoved replica from the config."""
+        if not self.state.db_initialised:
+            return
+        self.mongo_manager.process_unremoved_units()
+        if set(self.state.app_peer_data.replica_set_hosts) != self.state.app_hosts:
+            self.state.app_peer_data.replica_set_hosts = list(self.state.app_hosts)
+        self.update_related_hosts()
+
+    def update_related_hosts(self):
+        """Update the app relations that need to be made aware of the new set of hosts."""
+        if self.state.is_role(MongoDBRoles.REPLICATION):
+            for relation in self.state.client_relations:
+                self.mongo_manager.update_app_relation_data(relation)
+        # TODO: Update related hosts for config server , cluster.
+
+    def open_ports(self) -> None:
+        """Open ports on the workload.
+
+        VM-only.
+        """
+        if self.substrate != Substrates.VM:
+            return
+        ports = [MongoPorts.MONGODB_PORT]
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            ports.append(MongoPorts.MONGOS_PORT)
+
+        try:
+            for port in ports:
+                self.workload.exec(["open-port", f"{port}/TCP"])
+        except WorkloadExecError as e:
+            logger.exception(f"Failed to open port: {e}")
+            raise
+
+    @property
+    def primary(self) -> str | None:
+        """Retrieves the primary unit with the primary replica."""
+        with MongoConnection(self.state.mongo_config) as connection:
+            try:
+                primary_ip = connection.primary
+            except Exception as e:
+                logger.error(f"Unable to get primary: {e}")
+                return None
+
+        for unit in self.state.units:
+            if primary_ip == unit.host:
+                return unit.name
+        return None
+
+    @override
+    def start_charm_services(self):
+        """Start the relevant services.
+
+        If we are running as config-server, we should start both mongod and mongos.
+        """
+        self.workload.start()
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            self.mongos_workload.start()
+
+    @override
+    def stop_charm_services(self):
+        """Stop the relevant services.
+
+        If we are running as config-server, we should stop both mongod and mongos.
+        """
+        self.workload.stop()
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            self.mongos_workload.stop()
```

Comment:
should `mongos_workload` be stopped before `workload` ?

Reply:
This is copy pasted from former charms.
Maybe it's better to stop it before indeed.